commands/charm/pager.go

```diff
@@ -0,0 +1,217 @@
+		bufSize: 3 << (10 * 2), // 3MB
+	}
+	for _, opt := range opts {
+		opt(p)
+	}
+	var err error
+	p.buffer, err = circbuf.NewBuffer(p.bufSize)
+	if err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func PagerWithBufferSize(size int64) PagerOpt {
+	return func(p *Pager) {
+		p.bufSize = size
+	}
+}
+
+func PagerWithTitle(title string) PagerOpt {
+	return func(p *Pager) {
+		p.title = title
+	}
+}
+
+func (p *Pager) Write(b []byte) (int, error) {
+	n, err := p.buffer.Write(b)
+	p.prog.Send(msgUpdate{})
+	return n, err
+}
+
+func (p *Pager) Start(ctx context.Context) error {
+	p.model = newPagerModel(ctx, p.buffer, p.title)
+	prog := tea.NewProgram(
+		p.model,
+		tea.WithAltScreen(),       // use the full size of the terminal in its "alternate screen buffer"
+		tea.WithMouseCellMotion(), // turn on mouse support so we can track the mouse wheel
+	)
+	p.prog = prog
+
+	err := prog.Start()
+	fmt.Fprintln(Indent(4), p.buffer.String())
+	return err
+}
+
+type pagerModel struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	start    time.Time
+	title    string
+	buffer   WriterStringer
+	ready    bool
+	viewport viewport.Model
+}
+
+func newPagerModel(ctx context.Context, buffer WriterStringer, title string) *pagerModel {
+	m := &pagerModel{
+		buffer: buffer,
+		title:  title,
+		start:  time.Now(),
+	}
+	m.ctx, m.cancel = context.WithCancel(ctx)
+	return m
+}
+
+type msgQuit struct{}
+type msgUpdate struct{}
+type msgTick struct{}
+
+func (m *pagerModel) Init() tea.Cmd {
+	return tea.Batch(
+		m.timerTick(),
+		func() tea.Msg {
+			<-m.ctx.Done()
+			return msgQuit{}
+		},
+	)
+}
+
+func (m *pagerModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	var (
+		cmd  tea.Cmd
+		cmds []tea.Cmd
+	)
+
+	switch msg := msg.(type) {
+	case msgQuit:
+		return m, tea.Quit
+
+	case tea.KeyMsg:
+		if k := msg.String(); k == "ctrl+c" {
+			m.cancel()
+			return m, nil
+		}
+
+	case tea.WindowSizeMsg:
+		headerHeight := lipgloss.Height(m.headerView())
+		footerHeight := lipgloss.Height(m.footerView())
+		verticalMarginHeight := headerHeight + footerHeight
+
+		if !m.ready {
+			// Since this program is using the full size of the viewport we
+			// need to wait until we've received the window dimensions before
+			// we can initialize the viewport. The initial dimensions come in
+			// quickly, though asynchronously, which is why we wait for them
+			// here.
+			m.viewport = viewport.New(msg.Width, msg.Height-verticalMarginHeight)
+			m.viewport.YPosition = headerHeight
+			m.viewport.SetContent(m.buffer.String())
+			m.ready = true
+
+			// This is only necessary for high performance rendering, which in
+			// most cases you won't need.
+			//
+			// Render the viewport one line below the header.
+			m.viewport.YPosition = headerHeight + 1
+		} else {
+			m.viewport.Width = msg.Width
+			m.viewport.Height = msg.Height - verticalMarginHeight
+		}
+	case msgUpdate:
+		m.viewport.SetContent(m.buffer.String())
+		m.viewport.GotoBottom()
```

Comment:
Does this override mouse events? e.g. if I scroll up and new content comes in will it snap the view back to the bottom?

Reply:
Yes, that was intentional because right now there is no indication that new output was added if you're scrolled up. Ideally we'd either add a scrollbar or some badge at the bottom indicating the presence of unread logs